File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/s3/S3Storage.java
package uk.nhs.nhsx.core.aws.s3;

import org.apache.http.entity.ContentType;

import java.net.URL;
import java.util.Date;
import java.util.List;
import java.util.Optional;

public interface S3Storage {

    void upload(Locator locator, ContentType contentType, ByteArraySource bytes, List<MetaHeader> metaHeaders);

    default void upload(Locator locator, ContentType contentType, ByteArraySource bytes) {
        upload(locator, contentType, bytes, List.of());
    }

    Optional<URL> getSignedURL(Locator locator, Date expiration);

}

File: src/aws/lambdas/incremental_distribution/infra/src/main/kotlin/uk/nhs/nhsx/core/aws/s3/S3Storage.kt
package uk.nhs.nhsx.core.aws.s3

import uk.nhs.nhsx.core.ContentType
import java.net.URL
import java.util.*

interface S3Storage {
    fun upload(
        locator: Locator,
        contentType: ContentType,
        bytes: ByteArraySource,
        metaHeaders: List<MetaHeader>
    )

    fun upload(
        locator: Locator,
        contentType: ContentType,
        bytes: ByteArraySource
    ) {
        upload(locator, contentType, bytes, listOf())
    }

    fun getSignedURL(
        locator: Locator,
        expiration: Date
    ): Optional<URL>
}
